### PR TITLE
Switch to ES2017 target and ES module system

### LINKED ISSUE #8801: Investigate ESM support in bokehjs' build
Based on discussion in PR #8782.

> `TypeError: FlatBush is not a constructor`

The underlying problem is that flatbush uses an experimental module field in its package.json to define the ES module entry point, along with the standard main field. webpack and similar linkers seem to prefer module over main, despite its non-standard nature. I will need to do proper research on this subject to decide whether this is a good idea for us to support the module field or not (currently I'm leaning towards the later, due to non-standard by default nature). One can fix this issue on the client side by setting up the linker (assuming webpack) with:
```js
{ resolve: { mainFields: ['main', 'module'] } }
```
For rollup etc. similar configuration option should be hopefully possible.

### PR DESCRIPTION
**Pull request by mattpap**
This is an early preview. Previously I added support for generating ES6 (ES2015) bundles, but only for testing purpose. The goal of this PR is switch to an even more modern variant of ES, which is ES2017, and make that the default this time. The primary addition in ES2017 is native async/await support, which will be useful when getting bokehjs more asynchronous-capable (loading external resources, delayed rendering etc.). Making this the default, means that old browsers won't be supported anymore, which includes IE and phantomjs. For these, I intend to provide legacy bundles, which will be ES5-transpiled versions of the main bundles with polyfills included. This will be available behind a flag in bokeh's resources, so that if a user wants to support old browsers (e.g. for internal company purpose), then he will have state that explicitly when creating his visualizations. Using legacy bungles will come with both code size and speed penalties over time (the more modern features we will use, the heavier the legacy bundles will be). The alternative is to drop support now, which I think may be premature, though it would make testing and deployment much easier.

Given the nature of changes in this PR, I also switched bokehjs' modules from commonjs to ES modules. This will make bokehjs integrate more easily with tooling in the web ecosystem, which includes packaging with webpack, etc. This will also allow us to resolve issues with some of our dependencies, like flatbush or proj4 (which can finally be updated).

fixes #8801

### CODE REVIEW

**Comment by bryevdv** [comment]
@mattpap have you seen this? https://rollupjs.org/guide/en/ I was asked why we don't consider using this as it will apparently generate the necessary module format

**Comment by mattpap** [comment]
> have you seen this? https://rollupjs.org/guide/en/

Yes. Yet another bundler/linker with somewhat unique packing format. Not well suited for bokehjs.

> I was asked why we don't consider using this as it will apparently generate the necessary module format

I don't follow this statement (I mean I could infer what the author meant). A bundler/linker takes a set of modules as input and produces one or more bundles, which by themselves can be modules of some sort (AMD, CJS, ESM; UMD, global), not necessarily of the same type or connected to the original set of modules. This PR makes it easier for bokehjs' modules to work with a variety of modern bundlers (e.g. unlocking possibility of bundle optimizations; for what it's worth, especially given our requirements), what they produce is an orthogonal concern here.

File: bokeh/io/export.py [hunk 1]
```diff
@@ -30,7 +30,7 @@
 
 # Bokeh imports
 from ..embed import file_html
-from ..resources import INLINE
+from ..resources import LEGACY_INLINE
```
Comment (bryevdv) [inline bokeh/io/export.py]: Suggest `INLINE_LEGACY` or `INLINE_ES5`

File: bokeh/settings.py [hunk 2]
```diff
@@ -514,6 +514,10 @@ def dev(self):
     Whether Bokeh should use minified BokehJS resources.
     """)
 
+    legacy = PrioritizedSetting("legacy", "BOKEH_LEGACY", convert=convert_bool, default=False, dev_default=False, help="""
+    Whether Bokeh should use legacy (IE and phantomjs compatiblie) BokehJS resources.
```
Comment (bryevdv) [inline bokeh/settings.py]: compatiblie -> compatible

File: bokehjs/test/unit/models/glyphs/image_url.ts [hunk 3]
```diff
@@ -26,15 +26,16 @@ describe("ImageURL module", () => {
 
   describe("ImageURLView", () => {
 
-    it("`_set_data` should correctly set Image src", () => {
+    it.skip("`_set_data` should correctly set Image src", () => {
```
Comment (bryevdv) [inline bokehjs/test/unit/models/glyphs/image_url.ts]: is the test wrong, or?

Comment (mattpap) [reply bokehjs/test/unit/models/glyphs/image_url.ts]: The test was designed for being run under node, not in a browser, so now it needs to be "asynchronized", just there currently no good way to do this. I will restore this test when I will make bokeh more asynchronous capable.

File: bokeh/io/export.py (hunk 1 above)
Comment (mattpap) [reply bokeh/io/export.py]: I can rename it, but in any case, this is just a temporary measure to make exports' tests pass. I will be redesigning resources, and having separate definition for legacy inline resources most likely won't be necessary.

**Comment by mattpap** [comment]
Summing up all those changes, this PR accomplishes the following:

- bokehjs' bundles are now ES2017 compatible and there are ES5 legacy bundles available (have to be explicitly requested). Given that PR #9331 wasn't finished yet, `io.export` uses legacy bundles to make exports' work in `phantomjs`.
- bokehjs internally uses ES modules (ESM) instead of commonjs (CJS) module system. This affects how bokehjs is build, but doesn't have much bearing on how the bundles are consumed. Note there are still a few places where `require()` is used instead of imports. Most notably webgl bundle is loaded this way and all custom js (etc.) expose raw `require`. Both will fixed in future work. Also note that some of our dependencies are CJS-only, but that's a concern for a bundler.
- bokehjs' tests are now bundled and run in a web browser (chromium) using the same self-developed framework as bokehjs' integration tests use. There is no need anymore for jsdom, stubbing canvas, etc, and other hacks. Some of the stubs are still present and will be removed gradually. I had to temporarily disable connection unit tests, because they were designed to be run in a single process (similar though not exactly the same problem as we had with recent async/await switch in bokeh).

**Comment by bryevdv** [comment]
@mattpap selection tools are broken on this branch. On the selection histogram app there is a console error:

Screen Shot 2019-11-27 at 3 47 31 PM


On `color_sliders.py` it draws the overlay but makes no selection.

**Comment by bryevdv** [comment]
line10k webgl example looks like this:

Screen Shot 2019-11-27 at 4 13 07 PM

(usually works on osx)

**Comment by mattpap** [comment]
Apparently `TouchEvent` is not that commonly implemented. Previously it was guarded against specifically for jsdom. I suppose I will have to restore that code.

**Comment by mattpap** [comment]
> line10k webgl example looks like this:

Does it use webgl at all? Issue:
```
Bokeh.index[1001].renderer_views[1036].glyph.glglyph
```
Should be not null.

**Comment by mattpap** [comment]
GitHub CI now fails with unrelated errors both on Windows and OSX.

**Comment by mattpap** [comment]
> line10k webgl example looks like this:

I tested everything on Linux and Windows any don't see any WebGL-related issues, expect for it not working in IE, but that's a preexisting condition (did it ever worked?).

**Comment by bryevdv** [comment]
@mattpap This seems like maybe the downstream dependency changed somehow:

https://github.com/bokeh/bokeh/pull/9477/checks?check_run_id=325731006#step:8:515

**Comment by mattpap** [comment]
> This seems like maybe the downstream dependency changed somehow:

It looks like so, but I can't reproduce this on Linux. Also the Window CI job hangs on `conda install`. I tried testing in another PR and got the same outcome.

**Comment by bryevdv** [comment]
@mattpap 

* I'll try to figure out what dependency changed and we can pin it for now. 

* `selection_histogram.py` example almost works now. Selection is made and hist plots are updated, but the glyph styling on the actual scatter is never applied (the non-selected glyphs do not switch to appear non-selected)

* I'm not especially pressed about it, but yes the line10 webgl example did used to work on OSX

**Comment by mattpap** [comment]
> (...), but the glyph styling on the actual scatter is never applied

This is an issue in `master`. Regression was introduced in PR #9153.

**Comment by mattpap** [comment]
> (...) but yes the line10 webgl example did used to work on OSX

Can you test this on `master`?

**Comment by bryevdv** [comment]
@mattpap same result on Chrome and OS X:

ScreenFlow

Given the panning behavior it really seems like the clip region for just the gl line is being set based on an incorrect pixel ratio for this retina display.

**Pull request merged**